pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -132,7 +132,10 @@ func (configgen *ConfigGeneratorImpl) BuildListenerTLSContext(serverTLSSettings
 	alpnByTransport := util.ALPNHttp
 	if transportProtocol == istionetworking.TransportProtocolQUIC {
 		alpnByTransport = util.ALPNHttp3OverQUIC
+	} else if transportProtocol == istionetworking.TransportProtocolTCP {
+		alpnByTransport = util.ALPNDownstreamWithMxc
```

Comment:
What about HTTP

Reply:
Does it use pure http/1.1 for metadata exchange?  @howardjohn 

Looks we set "istio"

Reply:
for http metadata exchange, alpn is not checked. alpn is checked only in tcp MX case.

Reply:
Sorry I do not understand. For sidecar inbound listeners, we do NOT set this for HTTP types. This seems like a divergence?

Reply:
true. At sidecar inbound listeners(DownstreamTlsContext),  for HTTP we configure `h2, http/1.1` and for TCP we have `istio-peer-exchange, h2, http/1.1`, which means `istio-peer-exchange` is sent to clients only in the TCP case.
With this PR, ONLY IF ISTIO_MUTUAL,  gateways are sending `"istio-peer-exchange", "h2", "http/1.1"` in both the TCP and HTTP case to the downstream clients. Therefore you are correct on divergence because of additional `istio-peer-exchange` in HTTP case.
Sending `istio-peer-exchange` in HTTP case is harmless though because we are not exposing istio native alpns to the mesh external clients. IMHO, it should be ok.
If still we want to make gateways not to send `istio-peer-exchange` in HTTP case, this will require some interface changes. [BuildListenerTlsContext](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/gateway.go#L660) is not passed any hint today if server is HTTP or TLS or terminating-tcp. We will have to add that hint in the function arguments. Initially, as I said, it seemed harmless to me, wanted to avoid this churn.

Reply:
Due to the `istio.alpn` filter, the ALPN is actually overridden for HTTP. Its not `istio-peer-exchange`, and if we do send that it will break traffic on the receiving side.

Reply:
As discussed offline, true that due to alpn override, `istio-peer-exchange` will not be sent in the tls client hello by http clients, but this will not break the traffic. gateway(server) will send the negotiated alpn(one of the http1.1 or h2) in the server hello.
I tried locally the above explained behavior. Following are the wireshark captures of tls handshake:
**tls client hello from the http client:**
![image](https://user-images.githubusercontent.com/11936386/166390408-6f36eff7-6b36-4deb-a275-f554fb299d30.png)

**tls server hello from the gateway(istio-mtls server):**
![image](https://user-images.githubusercontent.com/11936386/166390500-6d190fa2-6b53-46f5-b047-5764a594dae9.png)

Note that listener on server side had istio-peer-exchange in the downstreamTlsContext during this experiment:
![image](https://user-images.githubusercontent.com/11936386/166390564-549fba5b-d245-47ec-a225-fbad6de33bdd.png)

If HCM is not there on the client side i.e no alpn override, in the that from the envoy cluster alpn configuration, `istio-peer-exchange` will be sent by client in tls client hello message. gateway server will respond to that with `istio-peer-exchange` and tcp metadata will get exchanged because of the `istio-peer-exchange` being the negotiated alpn.